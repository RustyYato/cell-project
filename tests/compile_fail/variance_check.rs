use cell_project::cell_project as cp;
use core::cell::Cell;

struct Foo<'a> {
    x: Option<&'a Cell<Foo<'a>>>,
}

impl<'a> Drop for Foo<'a> {
    fn drop(&mut self) {
        // `ourselves` is an &Cell<Self>.
        // NB: `Drop` is unsound.
        if let Some(ourselves) = self.x.as_ref() {
            // replace `self` (but this doesn't actually replace `self`)
            let is_x_none = ourselves.replace(Foo { x: None }).x.as_ref().is_none();
            // if we just moved out of `self`, and we had a `Some` originally,
            // how come this is a `None`?
            if is_x_none {
                println!("how did we get a None?");
            }
        }
    }
}

fn main() {
    let foo = Cell::new(Foo { x: None });
    let x = cp!(Foo<'_>, foo.x);
    x.set(Some(&foo));
}
