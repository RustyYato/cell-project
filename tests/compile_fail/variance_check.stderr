error[E0597]: `foo` does not live long enough
  --> tests/compile_fail/variance_check.rs:27:16
   |
27 |     x.set(Some(&foo));
   |                ^^^^ borrowed value does not live long enough
28 | }
   | -
   | |
   | `foo` dropped here while still borrowed
   | borrow might be used here, when `foo` is dropped and runs the destructor for type `Cell<Foo<'_>>`
